blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+the value last assigned to it (at run time). In our example, just before `f` is
+called, `x` is assigned the value `2`.
+
+Therefore, under dynamic scoping, the example program returns `2`.
+
+Dynamically scoped programming languages include bash and LaTeX. Emacs Lisp is
+dynamically scoped, but allows the programmer to opt into lexical scoping.
+Conversely, Perl and Common Lisp are lexically scoped by default but allow the
+programmer to choose dynamic scoping if desired.
+
+
+## What about R?
+
+The example is written in valid R. If you run it, you will see that the result
+is `1`. Therefore, according to our litmus test, **R is lexically scoped**. But
+is this the whole story?
+
+In _[Evaluating the Design of the R Language][mhov2012]_ (ECOOP 2012)
+([doi][mhov2012-doi]), Morandat, Hill, Osvald, and Vitek write:
+
+> As is often the case, R is lexically scoped up to the point it is not. R
+> is above all a dynamic language with full reflective access to the running
+> program’s data and representation.
+
+[mhov2012]: http://janvitek.org/pubs/ecoop12.pdf
+[mhov2012-doi]: https://doi.org/10.1007/978-3-642-31057-7_6
+
+
+### Circumventing scoping in R
+
+There are many ways to "defeat" scoping in R---made easier because it supports
+explicit manipulation of environments.
+
+Here are some different tricks we can try.
+
+#### Evaluating arbitrary code
+
+```r
+x <- 1
+f <- function(a) {
+  eval(parse(text = "x <- 0"))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 0
+```
+
+On line 3, we parse the string `"x <- 0"` into an unevaluated R expression and
+then evaluate it. This assigns `0` to `x`, which is no longer a free variable
+when it is used on line 4.
+
+#### Simulating dynamic scope
+
+```r
+x <- 1
+f <- function(a) {
+  get("x", envir = parent.frame()) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 2
+```
+
+On line 3, we perform an explicit variable lookup for `x`, but we do so in the
+environment `parent.frame()`, which refers to the calling function's
+environment.
+
+#### Constructing an arbitrary environment
+
+```r
+x <- 1
+f <- function(a) {
+  e <- new.env()
+  e$x <- 3
+  get("x", envir = e) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 3
+```
+
+On line 3, we construct an empty environment, `e`. Next, on line 4, we add a
+new binding to that environment, where `x` has the value `4`. Finally, on line
+5, we perform a variable lookup in environment `e`.
+
+#### Deleting bindings
+
+```r
+x <- 1
+f <- function(a) {
+  rm("x", envir = parent.env(environment()))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # Error in f(0) : object 'x' not found
+```
+
+On line 3, we delete the binding `x` from the given environment,
+`parent.env(environment())`. This refers to the parent (or enclosing)
+environment of the current environment of `f`, i.e. the environment where `f`
+was defined, which is the top-level environment.
+
+
+### What does "scoping in R is dynamic" mean?
```

Comment:
-1 for this section

Make the definitions clear at the start of this post. Then, if you want to discuss the paper here, give more context so we understand whether the authors are talking about the same definitions.

But if the point of this section is to say that other researchers are aware of the scope-bending features, it might be better to put a few links at the end of the post

Reply:
The point was to say "here is a paper that got it wrong, either because they misunderstood 'dynamic scope,' or because they used an ambiguous wording."

I'll add more context, something like "this paper describes how difficult it is to optimize R, and they say 'scoping is dynamic' without defining 'scope.' It's not clear if they mean 'dynamic scope' (which would be incorrect) or that scoping rules are evaluated at run time."